### ISSUE TITLE
Create REST controller for dashboard sharing

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#4481]
## Feature Description

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation Brief

* 

### Test Coverage

* 

## QA Brief

* 

## Changelog entry

*

**Comment by felixarntz** [comment]
@aaemnnosttv 
> that should include the complete module settings for each module where the owner has changed

Can you clarify that? You have `{settings, moduleSettings}`, but what would `settings` be here? Why do we need to return the entire module settings here? Also, what should the request data format be for this endpoint? I think the ACs here would be easier to follow if they included more specifics on exact request and response format shape.

**Comment by aaemnnosttv** [comment]
@felixarntz ACs are updated to clarify request and response 👍

**Comment by felixarntz** [comment]
@aaemnnosttv Thanks, that looks better. One of my questions from above I'd still like to clarify, why do we need to return the entire module settings here? Wouldn't it be sufficient to return only the owner IDs for each module after making changes? No other module setting should ever change from a data sharing request change, or am I missing something here?

**Comment by aaemnnosttv** [comment]
> One of my questions from above I'd still like to clarify, why do we need to return the entire module settings here? Wouldn't it be sufficient to return only the owner IDs for each module after making changes? No other module setting should ever change from a data sharing request change, or am I missing something here?

@felixarntz my thought was that settings could have changed in the meantime so this would be a good time to synchronize them. It was intended more as a safeguard than a requirement (we don't even need to return the owner ID since that would always be the current user's ID if we knew it was successful). I think this may be less relevant now that this would only affect all-admin modules which don't have service entities (the important settings to sync).

One thing to keep in mind however is that `receiveGetSettings` doesn't support a partial update to `savedSettings` which is what we'd want to update on the client – not just the current values which could be reset (unlikely in the current state of things). So maybe we should use the full settings here to properly update the saved values?

Generally speaking, if the owner changes in an API request, I think it makes sense to sync the module settings at the same time since the server is the source of truth for them. What do you think?

**Comment by felixarntz** [comment]
@aaemnnosttv That makes sense overall, however adjusting the endpoint schema to return the full settings although they can't really change as part of this still seems confusing to me. `settings` and `moduleSettings` are of substantially different shape, the first is a single setting group, the latter is one setting group _per_ module. That in combination with that the module settings can't change here, other than for the special `ownerID` makes me think we shouldn't return the entire module settings here.

Couldn't we only return the updated owner IDs in here, maybe even in a `newOwnerIDs` property keyed by module slug which is only present if it changed for a module? I think we would then be able to still pass everything to `receiveGetSettings` by getting the latest settings for the module, updating _just_ the `ownerID`, and then passing to `receiveGetSettings`. That still feels complicated, but at least only on the client-side instead of on the client-side _and_ the server-side.

**Comment by aaemnnosttv** [comment]
@felixarntz see my reply https://github.com/google/site-kit-wp/issues/4533#issuecomment-1000149024

One thing I just thought of is the problem this presents in terms of implementing using `createFetchStore` as we'd need to dispatch actions with the response but the `controlCallback` doesn't really allow for that, and the `reducerCallback` is of course limited to the current store. So in order for doing any of this we'd need to provide access to the registry in the `controlCallback`, or this updating would need to be handled outside of the fetch save settings action, which maybe the better option.

That's getting a bit off topic here but still relevant. I think returning `newOwnerIDs` SGTM as long as we had an action for updating only that part in state rather than pushing the current settings into saved. I'll update the ACs to use this approach instead.

**Comment by felixarntz** [comment]
@aaemnnosttv That sounds good to me. Let's eventually add an action like `refreshOwnerID` to the relevant module stores.

**Comment by asvinb** [comment]
@aaemnnosttv I started working on the IB for this and I have some questions:

- Can you let me know what shall the structure for the request be like? Is it the same as the response sent? If so, how will we know which setting is for which module?
- If we are updating multiple modules, does that mean the response will be an array containing settings for each module?

**Comment by aaemnnosttv** [comment]
> * Can you let me know what shall the structure for the request be like? Is it the same as the response sent? If so, how will we know which setting is for which module?
> * If we are updating multiple modules, does that mean the response will be an array containing settings for each module?

@asvinb, this is what this item from the AC describes

* the request should require a `data` parameter which should be an object of settings to merge into the current sharing settings similar to `modules/:slug/data/settings`

The settings would have the shape as defined in https://github.com/google/site-kit-wp/issues/4475 but the controller doesn't need to be particularly concerned with that here similar to how module settings are updated https://github.com/google/site-kit-wp/blob/06969446b8b18cec1d3bb4e0fdab35b08ae12dc4/includes/Core/Modules/Modules.php#L806

Sharing settings are grouped per-module but they're all stored in the same option – they aren't distributed to the individual module settings if maybe that's what you were thinking?


> * If we are updating multiple modules, does that mean the response will be an array containing settings for each module?

The response will only have 2 top-level keys: `settings` and `newOwnerIDs`.

The `settings` key of the response is simply the full module sharing settings, again similar to the modules’ settings endpoint only nested under the key rather being at the top-level of the response. https://github.com/google/site-kit-wp/blob/06969446b8b18cec1d3bb4e0fdab35b08ae12dc4/includes/Core/Modules/Modules.php#L808

The `newOwnerIDs` is a map of module slug -> new owner ID for any modules that changed ownership as a result of changing the sharing settings. This will only apply to Idea Hub and PSI for now but the implementation here doesn't need to be concerned with that either, it only needs to keep track of owner IDs that changed after updating the sharing settings.

I hope that makes more sense now. If not, I'm happy to jump on a call to explain!

**Comment by asvinb** [comment]
@aaemnnosttv Makes more sense. I've updated the IB, so let me know if that works. Else, let's jump on a call to iron things out 👍

**Comment by aaemnnosttv** [comment]
Thanks @asvinb !

> * The following file `includes/Core/Notifications/Notifications.php` can be taken as example for `REST_Dashboard_Sharing_Controller` with the following changes:

This class has some extra things in it which are specific to its needs for notifications. Maybe one of the other `REST_*_Controller` classes would be better to use as an example?

Otherwise this looks good to me, although I think we'll want to add a little more to the estimate to leave sufficient time for the tests 👍

**Comment by asvinb** [comment]
Thanks @aaemnnosttv . IB updated. Can you check again?

**Comment by aaemnnosttv** [comment]
Thanks @asvinb, this is very close but I realized there was an important detail missing in the ACs which I've added now regarding permissions. Apologies for that!

Also, I just opened #4620 which is what will be responsible for updating the module ownership when these settings are updated. This doesn't change anything for the IB here but I'm going to mark this issue as blocked by #4620 as the `newOwnerIDs` functionality depends on it.

**Comment by asvinb** [comment]
Sounds good @aaemnnosttv . IB updated to include more details about the `permission_callback` key.

**Comment by aaemnnosttv** [comment]
Thanks @asvinb ! There was another detail added at the same time about the updating of the settings related to capabilities (see the part about `googlesitekit_manage_module_sharing_options`), then this should be good to go.

**Comment by asvinb** [comment]
@aaemnnosttv IB updated.

**Comment by aaemnnosttv** [comment]
IB ✅

**Comment by FlicHollis** [comment]
Hi @jimmymadon 
As this is currently blocked by two tickets, one of which has not been started, I am going to move this out of Sprint 69 and to the backlog ready for Sprint 70. Let me know if you agree. 
Blockers are: 4620 (not yet started and blocked by 4579 in execution)
cc @eclarke1

**Comment by jimmymadon** [comment]
@FlicHollis Thanks - that makes sense. I have already executed most of the points in the IB so will keep this assigned to me.

**Comment by jimmymadon** [comment]
Wanted to clarify [why we auto-update the owner only for modules that implement the `Module_With_Service_Entity` interface.](Module_With_Service_Entity) Other modules can also have the `management` property set to `all_admins` which allow other admins to change sharing settings. When this happens, shouldn't the owner be automatically updated to the new admin who changed the settings last? 

So essentially, would it be safe for me to remove that check?

**Comment by aaemnnosttv** [comment]
> Wanted to clarify [why we auto-update the owner only for modules that implement the `Module_With_Service_Entity` interface.](https://github.com/google/site-kit-wp/blob/a6f85361d92845221586e86b826d13656b5c2444/includes/Core/Modules/Modules.php#L357) Other modules can also have the `management` property set to `all_admins` which allow other admins to change sharing settings. When this happens, shouldn't the owner be automatically updated to the new admin who changed the settings last?
> 
> So essentially, would it be safe for me to remove that check?

Just to follow up on the question here that was answered elsewhere –

See https://github.com/google/site-kit-wp/issues/4620

This handler is only for shared ownership modules (previously known as "all admin modules") which are those that don't have a service entity, which is why they're manageable by all admins. This handler updates the owner when the sharing settings change to be the current user.

For normal modules (with a service entity), the owner is the user who set up the connection in the module settings. Module ownership cannot be changed via dashboard sharing settings for these.

It's important to note that changing module ownership for a module with a service entity requires access to the service entity, otherwise the shared data would be broken (so we wouldn't want this behavior for such modules as it could easily cause problems). The management setting is there to allow other admins to manage who the module is shared with, without needing that access themselves.

For shared ownership modules, the access is more basic as there is no service level account the user making the API request needs access to, it just needs a token.

**Comment by jimmymadon** [comment]
@hussain-t pointed out that the issues which rely on this issue need the REST route to be registered. I originally thought this would be covered in a separate issue. So I have created a follow up PR that registers the route by initialising the controller in `Plugin.php`. I have also updated the QAB to now test this route thoroughly in the plugin. 

@aaemnnosttv  As per our discussion a few weeks back, this issue assumes that we will have some default dashboard sharing settings set before requests are made to this endpoint. This is because this endpoint will only merge settings _if_ the user has the `googlesitekit_manage_module_sharing_options($module_slug)` permission - the check for which requires the `$module_slug` to be present in the DB option (`googlesitekit_dashboard_sharing`).

cc. @techanvil

**Comment by aaemnnosttv** [comment]
> As per our discussion a few weeks back, this issue assumes that we will have some default dashboard sharing settings set before requests are made to this endpoint. This is because this endpoint will only merge settings _if_ the user has the `googlesitekit_manage_module_sharing_options($module_slug)` permission - the check for which requires the `$module_slug` to be present in the DB option (`googlesitekit_dashboard_sharing`).

@jimmymadon if the module slug wasn't present in the option, an empty array should be fine for the purpose of merging, no? If needed, we could add a method to the sharing settings class to get the value for a single module which includes defaults.

## ⚠️ 

On a separate note, I just noticed something in the implementation that should be corrected to match the ACs: https://github.com/google/site-kit-wp/blob/00a2fdf067e3931df0ba20a20f7df3ff1f8789a1/includes/Core/Modules/REST_Dashboard_Sharing_Controller.php#L97

Here the settings should be returned under the `settings` property. The ACs use ES6-like syntax for this but it should be `'settings' => $new_sharing_settings`

**Comment by jimmymadon** [comment]
@aaemnnosttv I have updated the PR to insert the default "all_admins" management sharing setting for shared ownership modules. So that works well now.

However, I further realised that currently, any admin can change the 'management' setting, as long as they have the `googlesitekit_manage_module_sharing_options` capability for that module. Is this correct? Should the `googlesitekit_delegate_module_sharing_management` capability be checked when modifying the 'management' setting?

**Comment by jimmymadon** [comment]
@eugene-manuilov I have addressed @aaemnnosttv's comments as per my comment above and have one question which could be done separately.

**Comment by eugene-manuilov** [comment]
PR looks good to me, but I would like @aaemnnosttv to look at it as well.

**Comment by aaemnnosttv** [comment]
@jimmymadon I'm still finishing the CR but noticed a few things in the QAB that could be enhanced

> * Useful `fetch` snippet if using the Chrome / browser console:

This snippet won't work because the WP nonce will be different for every user (looks like you know this already though from the instruction "replace the x-wp-nonce with a nonce from a recent report request on local"), but is also time-sensitive. Instead we can use `googlesitekit.api.(get|set)` respectively which will have the proper nonce configured automatically.

Regarding WP CLI commands to change the module owner/settings you can do this in a better "native" way instead:

```
wp option patch update googlesitekit_search-console_settings ownerID 2
```
See https://developer.wordpress.org/cli/commands/option/patch/ for the docs.

**Comment by aaemnnosttv** [comment]
@jimmymadon would you please update the QAB?

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- There are no console errors and everything works as before when setting up Site Kit and activating all Modules. For some additional testing, I checked that I was able to deactivate modules and also tested this on a site with two admin authorised users to ensure no issues occurred.  

Added the `QA:Eng` label for the required tests as per the QAB.

**Comment by hussain-t** [comment]
### QA Verified ✅ 
- Verified the new `sharing-settings` REST endpoint with the various scenarios mentioned in the QAB.
- Verified that the newly added PHPUnit tests are sufficient to cover the new functionality and pass.

**Comment by hussain-t** [comment]
### QA Update ⚠️ 

@jimmymadon, @aaemnnosttv, there is one observation I found later on that the endpoint returns the `newOwnerIDs` in the shape of an object like `{pagespeed-insights: 1}`. However, if there is no change in the `ownerID,` it returns an empty array. The shape of the `newOwnerIDs` property is an object with the data, and it's an array (empty) if there is no data. Do we need to return the `newOwnerIDs` if there is no change? WDYT?

**Comment by jimmymadon** [comment]
@aaemnnosttv Before sending the response here: https://github.com/google/site-kit-wp/blob/76cc88498c98ae65c4a4df13fd91f5baecf83d9b/includes/Core/Modules/REST_Dashboard_Sharing_Controller.php#L86-L100

I could add: 
```php
$changed_module_owners = empty( $changed_module_owners ) ? new stdClass() : $changed_module_owners;
```
to return an empty object {} OR just return null OR remove the `newOwnerIDs` key altogether. I think any of these responses could work for the front-end? I prefer the first option as it keeps the shape consistent?

**Comment by aaemnnosttv** [comment]
@hussain-t @jimmymadon This is a quirk of how PHP encodes an associative array to JSON. I think we just need to cast the `newOwnerIDs` to an object before returning from the controller so that the empty map doesn't get encoded as an array.

i.e. 
```
'newOwnerIDs' => (object) $changed_module_owners,
```

I'd prefer to keep the key in the response. We can always check `Object.keys( newOwnerIDs ).length` on the client to know if there are any or not.

**Comment by wpdarren** [comment]
### QA Update: ✅

**Verified:**

- There are no console errors and everything works as before when setting up Site Kit and activating all Modules.
- For some additional testing, I checked that I could activate all modules via settings and CTA on dashboard.
- I also checked that I was able to deactivate modules and also tested this on a site with two admin authorised users to ensure no issues occurred.

**Comment by wpdarren** [comment]
### QA Update: ✅

**Verified:**

- There are no console errors and everything works as before when setting up Site Kit and activating all Modules.
- For some additional testing, I checked that I could activate all modules via settings and CTA on dashboard.
- I also checked that I was able to deactivate modules and also tested this on a site with two admin authorised users to ensure no issues occurred.

**Comment by hussain-t** [comment]
### QA Verified ✅

- Verified the new sharing-settings REST endpoint with the various scenarios mentioned in the QAB.
- Verified that the newly added PHPUnit tests are sufficient to cover the new functionality and pass.
- **Verified the `newOwnerIDs` is an empty object if there are no changes instead of an empty array.**

<img width="826" alt="Screenshot 2022-05-19 at 10 01 34 PM" src="https://user-images.githubusercontent.com/24408261/169351971-b0d1ed77-dce6-450b-94cf-2dc56fa8aec0.png">
